From 7104d18ef58405681e1cc5096d1d7b35a79a6ea9 Mon Sep 17 00:00:00 2001
From: Cristian Papa <76759272+Cris7015@users.noreply.github.com>
Date: Wed, 29 Jul 2026 22:50:23 -0300
Subject: [PATCH] net: airoha: set EN751221 cascade and CPU-port IPG modes

The EN751221 integrated MT7530 uses port 5 for the TRGMII cascade
to the external switch and port 6 for the CPU link. The OEM setup
and working OpenWrt state use shrink IPG on P5 but short IPG on P6.

Program the two links separately instead of applying shrink IPG to
both.

Signed-off-by: Cristian Papa <76759272+Cris7015@users.noreply.github.com>
Signed-off-by: Matheus Sampaio Queiroga <srherobrine20@gmail.com>
---
 drivers/net/airoha_eth.c | 15 +++++++++++----
 1 file changed, 11 insertions(+), 4 deletions(-)

diff --git a/drivers/net/airoha_eth.c b/drivers/net/airoha_eth.c
index c3559f1e164..90ca408c48b 100644
--- a/drivers/net/airoha_eth.c
+++ b/drivers/net/airoha_eth.c
@@ -1157,13 +1157,20 @@ static int airoha_switch_init(struct udevice *dev, struct airoha_eth *eth)
 				 SWITCH_BC_FFP | SWITCH_UNM_FFP | SWITCH_UNU_FFP |
 				 SWITCH_CPU_EN | FIELD_PREP(SWITCH_CPU_PORT, 6));
 
-		pmcr = SWITCH_IPG_CFG_SHRINK | SWITCH_MAC_MODE |
-		       SWITCH_FORCE_MODE | SWITCH_MAC_TX_EN |
+		pmcr = SWITCH_MAC_MODE | SWITCH_FORCE_MODE | SWITCH_MAC_TX_EN |
 		       SWITCH_MAC_RX_EN | SWITCH_BKOFF_EN | SWITCH_BKPR_EN |
 		       SWITCH_FORCE_SPD_1000 | SWITCH_FORCE_DPX |
 		       SWITCH_FORCE_LNK;
-		airoha_switch_wr(eth, SWITCH_PMCR(5), pmcr);
-		airoha_switch_wr(eth, SWITCH_PMCR(6), pmcr);
+
+		/*
+		 * P5 is the TRGMII cascade to the external MCM switch while P6
+		 * is the CPU port.  The EN751221 OEM setup uses shrink IPG only
+		 * on the cascade and short IPG on the CPU link.
+		 */
+		airoha_switch_wr(eth, SWITCH_PMCR(5),
+				 pmcr | SWITCH_IPG_CFG_SHRINK);
+		airoha_switch_wr(eth, SWITCH_PMCR(6),
+				 pmcr | SWITCH_IPG_CFG_SHORT);
 		airoha_switch_wr(eth, SWITCH_PHY_POLL, 0x7f7f8c08);
 
 		return 0;
-- 
2.55.0

